src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs

```diff
@@ -184,7 +187,8 @@ private void ValidateService(ServiceDescriptor descriptor)
 
         internal void ReplaceServiceAccessor(ServiceCallSite callSite, Func<ServiceProviderEngineScope, object?> accessor)
         {
-            _realizedServices[callSite.ServiceType] = accessor;
+            // TODO BPETIT
+            _realizedServices[ServiceIdentifier.FromServiceType(callSite.ServiceType)] = accessor;
```

Comment:
This seems really bad. Doesn't this mean that a named services `Func<ServiceProviderEngineScope, object?>` could replace that of an equivalent non-named services. That could be really nasty to try to debug 😱. It also means that we always use the `CallSiteRuntimeResolver` and never upgrade to the `ILEmitResolverBuilder`.

This is also fixed by adding the name to the ServiceCallSite, right?

Reply:
Yes, looks like the ServiceIdentifier has no access to the service name so some ambiguity could be introduced. The updated and ServiceCallSite should be passed as parameter to the ServiceIdentifier method to return an indexer for the _realizedServiced collection.

Would the _realizedServices collection need to be updated as well?